File: inference/modelInference/sampleResnetAIPP/src/sampleResnetAIPP.cpp
#include <dirent.h>
#include <opencv2/opencv.hpp>
#include "acl/acl.h"
#include "label.h"
#define INFO_LOG(fmt, args...) fprintf(stdout, "[INFO]  " fmt "\n", ##args)
#define ERROR_LOG(fmt, args...) fprintf(stderr, "[ERROR]   " fmt "\n", ##args)

using namespace cv;
using namespace std;
typedef enum Result {
    SUCCESS = 0,
    FAILED = 1
} Result;

class SampleResnetAIPP {
    public:
    SampleResnetAIPP(int32_t device, const char* ModelPath, int32_t modelWidth, int32_t modelHeight);
    ~SampleResnetAIPP();
    Result InitResource();
    Result ProcessInput(const string testImgPath);
    Result Inference();
    Result GetResult();
    private:
    void ReleaseResource();
    int32_t deviceId_;
    aclrtContext context_;
    aclrtStream stream_;

    uint32_t modelId_;
    const char* modelPath_;
    int32_t modelWidth_;
    int32_t modelHeight_;
    aclmdlDesc *modelDesc_;
    aclmdlDataset *inputDataset_;
    aclmdlDataset *outputDataset_;
    void *picDevBuffer_;
    void* imageBuffer_;
    size_t devBufferSize_;
};

SampleResnetAIPP::SampleResnetAIPP(int32_t device, const char* modelPath,
                                   int32_t modelWidth, int32_t modelHeight) :
deviceId_(device), context_(nullptr), stream_(nullptr),
modelId_(0), modelPath_(modelPath), modelWidth_(modelWidth),
modelHeight_(modelHeight), modelDesc_(nullptr),
inputDataset_(nullptr), outputDataset_(nullptr),
picDevBuffer_(nullptr), imageBuffer_(nullptr)
{
}

SampleResnetAIPP::~SampleResnetAIPP()
{
    ReleaseResource();
}

Result SampleResnetAIPP::InitResource()
{
    // init acl resource
    const char *aclConfigPath = "";
    aclError ret = aclInit(aclConfigPath);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclInit failed, errorCode is %d", ret);
        return FAILED;
    }

    ret = aclrtSetDevice(deviceId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclrtSetDevice failed, errorCode is %d", ret);
        return FAILED;
    }

    ret = aclrtCreateContext(&context_, deviceId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclrtCreateContext failed, errorCode is %d", ret);
        return FAILED;
    }

    ret = aclrtCreateStream(&stream_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclrtCreateStream failed, errorCode is %d", ret);
        return FAILED;
    }

    // load model from file
    ret = aclmdlLoadFromFile(modelPath_, &modelId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclmdlLoadFromFile failed, errorCode is %d", ret);
        return FAILED;
    }

    // create description of model
    modelDesc_ = aclmdlCreateDesc();
    ret = aclmdlGetDesc(modelDesc_, modelId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclmdlGetDesc failed, errorCode is %d", ret);
        return FAILED;
    }

    // create data set of input
    int32_t index = 0;
    devBufferSize_ = aclmdlGetInputSizeByIndex(modelDesc_, index);
    aclrtMalloc(&picDevBuffer_, devBufferSize_, ACL_MEM_MALLOC_NORMAL_ONLY);
    inputDataset_ = aclmdlCreateDataset();
    aclDataBuffer *inputData = aclCreateDataBuffer(picDevBuffer_, devBufferSize_);
    ret = aclmdlAddDatasetBuffer(inputDataset_, inputData);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclmdlAddDatasetBuffer failed, errorCode is %d", ret);
        return FAILED;
    }

    // create data set of output
    outputDataset_ = aclmdlCreateDataset();
    size_t outputSize = aclmdlGetNumOutputs(modelDesc_);
    for (size_t i = 0; i < outputSize; ++i) {
        size_t modelOutputSize = aclmdlGetOutputSizeByIndex(modelDesc_, i);
        void *outputBuffer = nullptr;
        aclrtMalloc(&outputBuffer, modelOutputSize, ACL_MEM_MALLOC_NORMAL_ONLY);
        aclDataBuffer *outputData = aclCreateDataBuffer(outputBuffer, modelOutputSize);
        ret = aclmdlAddDatasetBuffer(outputDataset_, outputData);
        if (ret != ACL_SUCCESS) {
            ERROR_LOG("aclmdlAddDatasetBuffer failed, errorCode is %d", ret);
            return FAILED;
        }
    }
    return SUCCESS;
}

Result SampleResnetAIPP::ProcessInput(const string testImgPath)
{
    // read image from file by opencv
    Mat srcImage = imread(testImgPath);

    // zoom image to modelWidth_ * modelHeight_
    Mat resizedImage;
    resize(srcImage, resizedImage, Size(modelWidth_, modelHeight_));

    // covert image from BGR to RGB
    Mat rgbImage;
    cvtColor(resizedImage, rgbImage, CV_BGR2RGB);

    // pass image data to variable imageBuffer_
    imageBuffer_ = rgbImage.data;
    return SUCCESS;
}

Result SampleResnetAIPP::Inference()
{
    // copy image from imageBuffer_ to input data buffer
    aclError ret = aclrtMemcpy(picDevBuffer_, devBufferSize_,
                               imageBuffer_, devBufferSize_,
                               ACL_MEMCPY_HOST_TO_DEVICE);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("memcpy  failed, errorCode is %d", ret);
        return FAILED;
    }

    // inference
    ret = aclmdlExecute(modelId_, inputDataset_, outputDataset_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("execute model failed, errorCode is %d", ret);
        return FAILED;
    }
    return SUCCESS;
}

Result SampleResnetAIPP::GetResult() {
    // get result from output data set
    void *outHostData = nullptr;
    float *outData = nullptr;
    size_t outputIndex = 0;
    aclDataBuffer* dataBuffer = aclmdlGetDatasetBuffer(outputDataset_, outputIndex);
    void* data = aclGetDataBufferAddr(dataBuffer);
    uint32_t len = aclGetDataBufferSizeV2(dataBuffer);

    aclrtMallocHost(&outHostData, len);
    aclError ret = aclrtMemcpy(outHostData, len, data, len, ACL_MEMCPY_DEVICE_TO_HOST);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("memcpy  failed, errorCode is %d", ret);
        return FAILED;
    }
    outData = reinterpret_cast<float*>(outHostData);

    map<float, unsigned int, greater<float> > resultMap;
    for (unsigned int j = 0; j < len / sizeof(float); ++j) {
        resultMap[*outData] = j;
        outData++;
    }

    // do data processing with softmax and print top 5 classes
    double totalValue=0.0;
    for (auto it = resultMap.begin(); it != resultMap.end(); ++it) {
        totalValue += exp(it->first);
    }

    int cnt = 0;
    for (auto it = resultMap.begin(); it != resultMap.end(); ++it) {
        // print top 5
        if (++cnt > 5) {
            break;
        }
        INFO_LOG("top %d: index[%d] value[%lf] class[%s]", cnt, it->second,
                 exp(it->first) / totalValue, label[it->second].c_str());
    }

    ret = aclrtFreeHost(outHostData);
    outHostData = nullptr;
    outData = nullptr;
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclrtFreeHost failed, errorCode is %d", ret);
        return FAILED;
    }
    return SUCCESS;
}

void SampleResnetAIPP::ReleaseResource()
{
    aclError ret;
    imageBuffer_ = nullptr;
    picDevBuffer_ = nullptr;
    // release resource includes acl resource, data set and unload model
    for (size_t i = 0; i < aclmdlGetDatasetNumBuffers(inputDataset_); ++i) {
        aclDataBuffer *dataBuffer = aclmdlGetDatasetBuffer(inputDataset_, i);
        (void)aclDestroyDataBuffer(dataBuffer);
    }
    (void)aclmdlDestroyDataset(inputDataset_);
    inputDataset_ = nullptr;

    for (size_t i = 0; i < aclmdlGetDatasetNumBuffers(outputDataset_); ++i) {
        aclDataBuffer* dataBuffer = aclmdlGetDatasetBuffer(outputDataset_, i);
        void* data = aclGetDataBufferAddr(dataBuffer);
        (void)aclrtFree(data);
        (void)aclDestroyDataBuffer(dataBuffer);
    }
    (void)aclmdlDestroyDataset(outputDataset_);
    outputDataset_ = nullptr;

    ret = aclmdlDestroyDesc(modelDesc_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("destroy description failed, errorCode is %d", ret);
    }

    ret = aclmdlUnload(modelId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("unload model failed, errorCode is %d", ret);
    }

    if (stream_ != nullptr) {
        ret = aclrtDestroyStream(stream_);
        if (ret != ACL_SUCCESS) {
            ERROR_LOG("aclrtDestroyStream failed, errorCode is %d", ret);
        }
        stream_ = nullptr;
    }

    if (context_ != nullptr) {
        ret = aclrtDestroyContext(context_);
        if (ret != ACL_SUCCESS) {
            ERROR_LOG("aclrtDestroyContext failed, errorCode is %d", ret);
        }
        context_ = nullptr;
    }

    ret = aclrtResetDevice(deviceId_);
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclrtResetDevice failed, errorCode is %d", ret);
    }

    ret = aclFinalize();
    if (ret != ACL_SUCCESS) {
        ERROR_LOG("aclFinalize failed, errorCode is %d", ret);
    }
}

int main()
{
    const char* modelPath = "../model/resnet50.om";
    const string imagePath = "../data";
    int32_t device = 0;
    int32_t modelWidth = 224;
    int32_t modelHeight = 224;

    // all images in dir
    DIR *dir = opendir(imagePath.c_str());
    if (dir == nullptr)
    {
         ERROR_LOG("file folder does no exist, please create folder %s", imagePath.c_str());
        return FAILED;
    }
    vector<string> allPath;
    struct dirent *entry;
    while ((entry = readdir(dir)) != nullptr)
    {
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0
        || strcmp(entry->d_name, ".keep") == 0)
        {
            continue;
        }else{
            string name = entry->d_name;
            string imgDir = imagePath + "/" + name;
            allPath.push_back(imgDir);
        }
    }
    closedir(dir);

    if (allPath.size() == 0){
         ERROR_LOG("the directory is empty, please download image to %s", imagePath.c_str());
        return FAILED;
    }

    string fileName;
    SampleResnetAIPP sampleResnet(device, modelPath, modelWidth, modelHeight);
    Result ret = sampleResnet.InitResource();
    if (ret != SUCCESS) {
        ERROR_LOG("InitResource  failed");
        return FAILED;
    }

    for (size_t i = 0; i < allPath.size(); i++)
    {
        fileName = allPath.at(i).c_str();

        ret = sampleResnet.ProcessInput(fileName);
        if (ret != SUCCESS) {
            ERROR_LOG("ProcessInput  failed");
            return FAILED;
        }

        ret = sampleResnet.Inference();
        if (ret != SUCCESS) {
            ERROR_LOG("Inference  failed");
            return FAILED;
        }

        ret = sampleResnet.GetResult();
        if (ret != SUCCESS) {
            ERROR_LOG("GetResult  failed");
            return FAILED;
        }
    }
    return SUCCESS;
}

File: inference/modelInference/sampleResnetAIPP/python/src/sampleResnetAIPP.py
import os
import numpy as np
import cv2
import acl
from label import label

ACL_MEM_MALLOC_HUGE_FIRST = 0
ACL_MEMCPY_HOST_TO_DEVICE = 1
ACL_MEMCPY_DEVICE_TO_HOST = 2
NPY_FLOAT32 = 11
ACL_SUCCESS = 0

class sample_resnet_aipp(object):
    def __init__(self, device_id, model_path):
        self.device_id = device_id      # int
        self.context = None             # pointer
        self.stream = None
        
        self.model_id = None            # pointer
        self.model_path = model_path    # string 
        self.model_desc = None          # pointer when using

        self.input_dataset = None
        self.output_dataset = None

    def init_resource(self):
        # init acl resource
        ret = acl.init()
        if ret != ACL_SUCCESS:
            print('acl init failed, errorCode is', ret)
        ret = acl.rt.set_device(self.device_id)
        if ret != ACL_SUCCESS:
            print('set device failed, errorCode is', ret)
        self.context, ret = acl.rt.create_context(self.device_id)
        if ret != ACL_SUCCESS:
            print('create context failed, errorCode is', ret)
        self.stream, ret = acl.rt.create_stream()
        if ret != ACL_SUCCESS:
            print('create stream failed, errorCode is', ret)

        # load model from file
        self.model_id, ret = acl.mdl.load_from_file(self.model_path)

        # create model description
        self.model_desc = acl.mdl.create_desc()
        ret = acl.mdl.get_desc(self.model_desc, self.model_id)
        if ret != ACL_SUCCESS:
            print('get desc failed, errorCode is', ret)

        # create input data set
        input_size = acl.mdl.get_num_inputs(self.model_desc)
        self.input_dataset = acl.mdl.create_dataset()
        for i in range(input_size):
            input_buffer_size = acl.mdl.get_input_size_by_index(self.model_desc, i)
            input_buffer, ret = acl.rt.malloc(input_buffer_size,
                                              ACL_MEM_MALLOC_HUGE_FIRST)
            if ret != ACL_SUCCESS:
                print('create input_buffer failed, errorCode is', ret)
            input_data_buffer = acl.create_data_buffer(input_buffer, input_buffer_size)
            acl.mdl.add_dataset_buffer(self.input_dataset, input_data_buffer)

        # create output data set
        output_size = acl.mdl.get_num_outputs(self.model_desc)
        self.output_dataset = acl.mdl.create_dataset()
        for i in range(output_size):
            output_buffer_size = acl.mdl.get_output_size_by_index(self.model_desc, i)
            output_buffer, ret = acl.rt.malloc(output_buffer_size,
                                               ACL_MEM_MALLOC_HUGE_FIRST)

            output_data_buffer = acl.create_data_buffer(output_buffer, output_buffer_size)
            acl.mdl.add_dataset_buffer(self.output_dataset, output_data_buffer)

    def process_input(self, input_path, model_width, model_height):
        # read image from file by opencv
        input_path = os.path.abspath(input_path)
        image = cv2.imread(input_path)

        # zoom image to model_width * model_height
        resize_image = cv2.resize(image, (model_width, model_height))

        # switch channel BGR to RGB
        RGB_image = cv2.cvtColor(resize_image, cv2.COLOR_BGR2RGB)
        RGB_image = RGB_image.astype("uint8")
        result = np.frombuffer(RGB_image.tobytes(), np.uint8)
        images = [result]

        # pass image data to input data buffer
        policy = ACL_MEMCPY_HOST_TO_DEVICE
        input_num = acl.mdl.get_dataset_num_buffers(self.input_dataset)
        for i in range(input_num):
            if "bytes_to_ptr" in dir(acl.util):
                bytes_data = images[i].tobytes()
                ptr = acl.util.bytes_to_ptr(bytes_data)
            else:
                ptr = acl.util.numpy_to_ptr(images[i])
            input_data_buffer = acl.mdl.get_dataset_buffer(self.input_dataset, i)
            input_data_buffer_addr = acl.get_data_buffer_addr(input_data_buffer)
            input_data_size = acl.get_data_buffer_size(input_data_buffer)
            ret = acl.rt.memcpy(input_data_buffer_addr,
                                input_data_size,
                                ptr,
                                input_data_size,
                                policy)
            if ret != ACL_SUCCESS:
                print('memcpy failed, errorCode is', ret)

    def inference(self):
        # inference
        ret = acl.mdl.execute(self.model_id,
                              self.input_dataset,
                              self.output_dataset)
        if ret != ACL_SUCCESS:
            print('execute model failed, errorCode is', ret)

    def get_result(self):
        # get result from output data set
        res = []
        output_index = 0
        output_data_buffer = acl.mdl.get_dataset_buffer(self.output_dataset, output_index)
        output_data_buffer_addr = acl.get_data_buffer_addr(output_data_buffer)
        output_data_size = acl.get_data_buffer_size(output_data_buffer)
        ptr, ret = acl.rt.malloc_host(output_data_size)
        res.append({"size": output_data_size, "buffer": ptr})
        ret = acl.rt.memcpy(ptr,
                            output_data_size,
                            output_data_buffer_addr,
                            output_data_size,
                            ACL_MEMCPY_DEVICE_TO_HOST)
        if ret != ACL_SUCCESS:
            print('memcpy failed, errorCode is', ret)

        result = []
        out_index = 0
        dims, ret = acl.mdl.get_cur_output_dims(self.model_desc, out_index)
        out_dim = dims['dims']
        for temp in res:
            ptr = temp["buffer"]
            if "ptr_to_bytes" in dir(acl.util):
                bytes_data = acl.util.ptr_to_bytes(ptr, temp["size"])
                data = np.frombuffer(bytes_data, dtype=np.float32).reshape(tuple(out_dim))
            else:
                data = acl.util.ptr_to_numpy(ptr, tuple(out_dim), NPY_FLOAT32)
            result.append(data)

        # do data processing with softmax
        vals = np.array(result).flatten()
        vals = np.exp(vals)/np.sum(np.exp(vals))

        # print top 5 classes
        top_k = vals.argsort()[-1:-6:-1]
        print("======== top5 inference results: =============")
        for j in top_k:
            label_class = ""
            label_string = label.get(str(j))
            if label_string:
                list_iterator = iter(range(len(label_string)))
                for i in list_iterator:
                    if i == len(label_string) - 1:
                        label_class += label_string[i]
                else:
                    label_class += label_string[i] + "," + "\t"
                print("label:%d confidence:%f class:%s" % (j, vals[j], label_class))
            else:
                raise Exception("the key of label is not exist")

        for item in res:
            ptr = item['buffer']
            ret = acl.rt.free_host(ptr)
            if ret != ACL_SUCCESS:
                print('free host failed, errorCode is', ret)

    def release_resource(self):
        # release resource includes acl resource, data set and unload model
        for dataset in [self.input_dataset, self.output_dataset]:
            if not dataset:
                continue
            number = acl.mdl.get_dataset_num_buffers(dataset)
            for i in range(number):
                data_buf = acl.mdl.get_dataset_buffer(dataset, i)
                if data_buf:
                    acl.destroy_data_buffer(data_buf)
            acl.mdl.destroy_dataset(dataset)
        ret = acl.mdl.unload(self.model_id)
        if ret != ACL_SUCCESS:
            print('unload model failed, errorCode is', ret)
        if self.model_desc:
            acl.mdl.destroy_desc(self.model_desc)
            self.model_desc = None

        if self.context:
            ret = acl.rt.destroy_context(self.context)
            if ret != ACL_SUCCESS:
                print('destroy context failed, errorCode is', ret)
            self.context = None

        ret = acl.rt.reset_device(self.device_id)
        if ret != ACL_SUCCESS:
            print('reset device failed, errorCode is', ret)
        ret = acl.finalize()
        if ret != ACL_SUCCESS:
            print('finalize failed, errorCode is', ret)

if __name__ == '__main__':
    device = 0
    model_width = 224
    model_height = 224
    current_dir = os.path.dirname(os.path.abspath(__file__))
    model_path = os.path.join(current_dir, "../model/resnet50.om")
    if not os.path.exists(model_path):
        raise Exception("the model is not exist")

    # read all file path of image
    images_path = os.path.join(current_dir, "../data")
    if not os.path.exists(images_path):
        raise Exception("the directory is not exist")
    all_path = []
    for path in os.listdir(images_path):
        if path != '.keep':
            total_path = os.path.join(images_path, path)
            all_path.append(total_path)
    if len(all_path) == 0:
        raise Exception("the directory is empty, please download image")

    # inference
    net = sample_resnet_aipp(device, model_path)
    net.init_resource()
    for image in all_path:
        net.process_input(image, model_width, model_height)
        net.inference()
        net.get_result()
    print("*****run finish******")
    net.release_resource()
